cross-compiling for ARM

I'm considering cross-compilation tensorflow_cc for ARM arch. I tried https://github.com/lhelontra/tensorflow-on-arm and it works OK, but it uses Bazel and builds shared library only, while static one would be better for me, and I do not need python  packages. I do not like Bazel and cmake looks better for me.

So my question is how to cross-compile tensorflow_cc for ARM, I'm using https://github.com/dockcross/dockcross docker images for that.

Please advice what needs to be changed in tensorflow_cc for cross-compiling - probably CC_OPT_FLAGS, anything else? I'm going to set CC options per tensorflow-on-arm config for my arch (Beaglebone Black) https://github.com/lhelontra/tensorflow-on-arm/blob/master/build_tensorflow/configs/beagle_black.conf#L26 - will it be enough to enable TF optimization for my CPU (NEON SIMD in particular)? I'm new with TF, so I'm not sure if TF ARM toolchain may enable something special in Bazel build.

Thanks in advance!

Please disregard my previous comment - it was a broken curl executable in my docker image derived from dockcross one. 

Now I got further:
```
checking whether we are cross compiling... configure: error: in `/work/tensorflow_cc/build/tensorflow/tensorflow/contrib/makefile/downloads/protobuf':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
CMakeFiles/tensorflow_static.dir/build.make:106: recipe for target 'tensorflow-stamp/tensorflow_static-configure' failed
make[2]: *** [tensorflow-stamp/tensorflow_static-configure] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/tensorflow_static.dir/all' failed
```

--- a/Dockerfiles/install-common.sh
+++ b/Dockerfiles/install-common.sh
@@ -24,6 +24,7 @@ make
 rm -rf /home/tensorflow_cc/.cache
 rm -rf /root/.cache
 make install
+ldconfig
 cd "$cwd"
 rm -rf tensorflow_cc/tensorflow_cc/build
 

--- a/README.md
+++ b/README.md
@@ -83,6 +83,7 @@ mkdir build && cd build
 cmake ..
 make
 sudo make install
+sudo ldconfig
 ```
 
 **Warning:** Optimizations for Intel CPU generation `>=haswell` are enabled by default. If you have a

--- a/tensorflow_cc/CMakeLists.txt
+++ b/tensorflow_cc/CMakeLists.txt
@@ -52,7 +52,7 @@ target_include_directories(
 )
 target_link_libraries(
   tensorflow_cc INTERFACE
-  $<INSTALL_INTERFACE:lib/libtensorflow_cc.so.${PROJECT_VERSION_MAJOR}>
+  "${CMAKE_INSTALL_PREFIX}/lib/libtensorflow_cc.so.${PROJECT_VERSION_MAJOR}"
   dl pthread
 )
 

--- a/tensorflow_cc/CMakeLists.txt
+++ b/tensorflow_cc/CMakeLists.txt
@@ -45,14 +45,14 @@ target_compile_features(tensorflow_cc INTERFACE ${TARGET_CXX_STANDARD})
 # The include folders are sometimes contained under bazel-bin/bin/ and sometimes just bazel-bin.
 target_include_directories(
   tensorflow_cc INTERFACE
-  "${CMAKE_INSTALL_PREFIX}/include/tensorflow/bazel-bin/tensorflow/include"
-  "${CMAKE_INSTALL_PREFIX}/include/tensorflow/bazel-bin/tensorflow/include/src"
-  "${CMAKE_INSTALL_PREFIX}/include/tensorflow/bazel-bin/bin/tensorflow/include"
-  "${CMAKE_INSTALL_PREFIX}/include/tensorflow/bazel-bin/bin/tensorflow/include/src"
+  $<INSTALL_INTERFACE:include/tensorflow/bazel-bin/tensorflow/include>
+  $<INSTALL_INTERFACE:include/tensorflow/bazel-bin/tensorflow/include/src>
+  $<INSTALL_INTERFACE:include/tensorflow/bazel-bin/bin/tensorflow/include>
+  $<INSTALL_INTERFACE:include/tensorflow/bazel-bin/bin/tensorflow/include/src>
 )
 target_link_libraries(
   tensorflow_cc INTERFACE
-  "${CMAKE_INSTALL_PREFIX}/lib/libtensorflow_cc.so.${PROJECT_VERSION_MAJOR}"
+  $<INSTALL_INTERFACE:lib/libtensorflow_cc.so.${PROJECT_VERSION_MAJOR}>
   dl pthread
 )
 

--- a/tensorflow_cc/CMakeLists.txt
+++ b/tensorflow_cc/CMakeLists.txt
@@ -1,9 +1,10 @@
 cmake_minimum_required(VERSION 3.3 FATAL_ERROR)
+cmake_policy(SET CMP0048 NEW)  # Enable version parameter in project().
 file(STRINGS "${CMAKE_CURRENT_SOURCE_DIR}/PROJECT_VERSION" version)
 project(
   "tensorflow_cc"
+  VERSION ${version}
 )
-set(PROJECT_VERSION "${version}")
 
 # If enabled, bazel has to be installed.
 option(ALLOW_CUDA "Try to find and use CUDA." ON)

--- a/tensorflow_cc/cmake/TensorflowBase.cmake
+++ b/tensorflow_cc/cmake/TensorflowBase.cmake
@@ -19,4 +19,7 @@ ExternalProject_Add(
           COMMAND mkdir -p "${CMAKE_CURRENT_BINARY_DIR}/tensorflow/bazel-bin/bin/tensorflow/include/src"
           COMMAND touch "${CMAKE_CURRENT_BINARY_DIR}/tensorflow/bazel-bin/tensorflow/include/src/__placeholder__.h"
           COMMAND touch "${CMAKE_CURRENT_BINARY_DIR}/tensorflow/bazel-bin/bin/tensorflow/include/src/__placeholder__.h"
+  # Fix compilation with CUDA 11.3.
+  PATCH_COMMAND cp "${CMAKE_CURRENT_SOURCE_DIR}/patches/v2.5.0-Fix-NCCL-build-failure-with-CUDA-11.3.patch" .
+        COMMAND patch -p1 < v2.5.0-Fix-NCCL-build-failure-with-CUDA-11.3.patch
 )

--- a/tensorflow_cc/cmake/build_tensorflow.sh.in
+++ b/tensorflow_cc/cmake/build_tensorflow.sh.in
@@ -61,9 +61,7 @@ if [ "$cuda_allowed" == true ] && [ "$cuda_available" == true ]; then
 
     # choose the right version of CUDA compiler
     if [ -z "$GCC_HOST_COMPILER_PATH" ]; then
-        if   hash gcc 2>/dev/null && version_gt 10.3 `gcc -dumpversion`; then
-            export GCC_HOST_COMPILER_PATH=${GCC_HOST_COMPILER_PATH:-"/usr/bin/gcc"}
-        elif hash gcc-10 2>/dev/null && version_gt 10.3 `gcc-10 -dumpversion`; then
+        if hash gcc-10 2>/dev/null && version_gt 10.3 `gcc-10 -dumpversion`; then
             export GCC_HOST_COMPILER_PATH=${GCC_HOST_COMPILER_PATH:-"/usr/bin/gcc-10"}
         elif hash gcc-9 2>/dev/null && version_gt 9.4 `gcc-9 -dumpversion`; then
             export GCC_HOST_COMPILER_PATH=${GCC_HOST_COMPILER_PATH:-"/usr/bin/gcc-9"}

--- a/tensorflow_cc/patches/v2.5.0-Fix-NCCL-build-failure-with-CUDA-11.3.patch
+++ b/tensorflow_cc/patches/v2.5.0-Fix-NCCL-build-failure-with-CUDA-11.3.patch
@@ -1,0 +1,13 @@
+diff --git a/third_party/nccl/build_defs.bzl.tpl b/third_party/nccl/build_defs.bzl.tpl
+index ceb65e3b632f3..c875717403857 100644
+--- a/third_party/nccl/build_defs.bzl.tpl
++++ b/third_party/nccl/build_defs.bzl.tpl
+@@ -204,7 +204,7 @@ def _prune_relocatable_code_impl(ctx):
+             arguments = arguments,
+             mnemonic = "nvprune",
+         )
+-        output.append(outputs)
++        outputs.append(output)
+ 
+     return DefaultInfo(files = depset(outputs))
+ 

